File: julia/foldl.jl
sections = [
  Dict(
    "title" => "Getting started",
    "reset_lesson_position" => false,
    "lessons" => [
      Dict("name" => "Welcome"),
      Dict("name" => "Installation")
    ]
  ),
  Dict(
    "title" => "Basic operator",
    "reset_lesson_position" => false,
    "lessons" => [
      Dict("name" => "Addition / Subtraction"),
      Dict("name" => "Multiplication / Division")
    ]
  ),
  Dict(
    "title" => "Advanced topics",
    "reset_lesson_position" => true,
    "lessons" => [
      Dict("name" => "Mutability"),
      Dict("name" => "Immutability")
    ]
  )
]

struct Lesson
  name::String
  position::Int
end

struct Section
  title::String
  reset_lesson_position::Bool
  position::Int
  lessons::Vector{Lesson}
end

traversed_sections, _ = foldl(enumerate(sections); init=([], 0)) do (traversed_sections, lesson_counter), (section_pos, section)
  reset_lesson_position = section["reset_lesson_position"]

  if reset_lesson_position
    lesson_counter = 0
  end

  new_section = Section(
    section["title"],
    reset_lesson_position,
    section_pos,
    map(enumerate(section["lessons"])) do (lesson_pos, lesson)
      Lesson(lesson["name"], lesson_counter + lesson_pos)
    end
  )

  (
    [traversed_sections..., new_section],
    lesson_counter + length(new_section.lessons)
  )
end

dump(traversed_sections)

File: elixir/foldl.exs
defmodule FolderSolver do
  @moduledoc """
  With assumption that everything is a List, we only need to fold it right!

  As we also need to track additional data (section or lesson positions)
  between the sections, we use a two elements tuple to pass these pieces of
  information between the folds.

  Though this little tuple-hack makes us to use `elem/2` before we can return
  a true output result (as we have also to reverse accumulated lists).

  The similar approach with tuple-based accumulator we use for the lessons
  folding flow.
  """

  def solve(input) do
    input
    |> List.foldl({[], 1}, &fold_section/2)
    |> elem(0)
    |> Enum.reverse()
  end

  defp fold_section(section, {sections, lesson_position}) do
    # reset if needed, but pass the previous value if not
    lesson_position =
      case section[:reset_lesson_position] do
        true -> 1
        _ -> lesson_position
      end

    lessons =
      section[:lessons]
      |> List.foldl({[], lesson_position}, &fold_lesson/2)
      |> elem(0)
      |> Enum.reverse()

    section_position = length(sections) + 1

    section_with_position =
      section
      |> Map.put(:position, section_position)
      |> Map.put(:lessons, lessons)

    sections = [section_with_position | sections]
    lesson_position = lesson_position + length(lessons)

    {sections, lesson_position}
  end

  defp fold_lesson(lesson, {lessons, lesson_position}) do
    lesson_with_position =
      lesson
      |> Map.put(:position, lesson_position)

    {[lesson_with_position | lessons], lesson_position + 1}
  end
end

#########################
### RUN, FOREST, RUN! ###
#########################

input = [
  %{
    title: "Getting started",
    reset_lesson_position: false,
    lessons: [
      %{name: "Welcome"},
      %{name: "Installation"}
    ]
  },
  %{
    title: "Basic operator",
    reset_lesson_position: false,
    lessons: [
      %{name: "Addition / Subtraction"},
      %{name: "Multiplication / Division"}
    ]
  },
  %{
    title: "Advanced topics",
    reset_lesson_position: true,
    lessons: [
      %{name: "Mutability"},
      %{name: "Immutability"}
    ]
  }
]

FolderSolver.solve(input)
|> IO.inspect()
